File: backend/src/mellow/providers/knowledge.py
"""知识库抽象接口。

所有知识库实现必须遵循此协议。
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field


@dataclass
class WordEntry:
    """词条查询结果。"""
    word: str
    part_of_speech: str | None = None
    phonetic: str | None = None
    definitions: list[str] = field(default_factory=list)
    examples: list[str] = field(default_factory=list)
    synonyms: list[str] = field(default_factory=list)
    source: str = "unknown"


@dataclass
class SearchResult:
    """语义检索结果。"""
    content: str
    score: float = 0.0
    metadata: dict = field(default_factory=dict)


class KnowledgeProvider(ABC):
    """知识库提供者抽象接口。

    支持两种检索模式：
    - lookup(): 精确查词，用于学习计划，必须准确
    - search(): 语义检索，用于日常对话，可容错
    """

    @abstractmethod
    async def lookup(self, word: str) -> WordEntry | None:
        """精确查词 —— 返回标准化词条，未找到返回 None。"""
        ...

    @abstractmethod
    async def search(self, query: str, top_k: int = 5) -> list[SearchResult]:
        """语义检索 —— 返回相关知识点列表。"""
        ...

    @property
    @abstractmethod
    def source_name(self) -> str:
        """知识来源标识，用于溯源和展示。"""
        ...

File: frontend/src/api/knowledge.ts
/**
 * 知识库 API — 查词、语义搜索
 */

import { api } from './client';

// ===== Types =====

export interface WordEntry {
  word: string;
  phonetic: string | null;
  part_of_speech: string | null;
  definitions: string[];
  examples: string[];
  synonyms: string[];
  source: string;
}

export interface SearchResultItem {
  content: string;
  score: number;
  metadata: Record<string, unknown>;
}

export interface SearchResponse {
  query: string;
  results: SearchResultItem[];
}

// ===== API 函数 =====

export async function lookupWord(word: string): Promise<WordEntry> {
  return api.get<WordEntry>('/api/v1/knowledge/lookup', {
    params: { word },
  });
}

export async function searchKnowledge(q: string, topK: number = 5): Promise<SearchResponse> {
  return api.get<SearchResponse>('/api/v1/knowledge/search', {
    params: { q, top_k: topK },
  });
}